### ISSUE TITLE
Reactive stream processing

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#779]
# Current State

The current implementation of the stream processing consist of one StreamProcessController per StreamProcessor and each StreamProcessor can have a lot of EventProcessors. Besides that, each StreamProcessController has his own BufferedLogStreamReader and LogStreamWriter. 

The StreamProcesssors are not interested in all events, they can define a filter for the specific wanted EventTypes. The StreamProcessController have to read each event
if the filter of the StreamProcessor applies the StreamProcessor#onEvent method is called. The processor deserializes the LoggedEvent and returns an specific EventProcessor which should be used for this kind of Event.

# Idea

Since the most of the StreamProcessors are on the same position it is not really necessary that each processor has his own reader. Furthermore they are not interested in every event, but they have to read and deserialize each event to decide if they want to process it. 

The Idea is to create an ReactiveStreamProcessor, which reads each event of the LogStream, deserializes this Event into the correct type and distribute a copy of the Event to the StreamProcessors, which are interested in this kind of event. This means it should be possible to register an StreamProcessor for an EventType (or later for an EventState like CREATE, CREATED etc.). 

Note: The indirection with StreamProcessors and EventProcessors can be simplified
with the following issue https://github.com/zeebe-io/zeebe/issues/367 
If we have the Event state in the headers we can remove the StreamProcessors and only have EventProcessors, which makes the code cleaner. 

The benefit of this approach would be that we have only one that reads and deserialize the events. This would save a lot of processing and IO time, besides that we don't need to reinvent the wheel in each StreamProcessor, since they would got the complete deserialized event (like an TaskEvent). Before, each StreamProcessor got an LoggedEvent and had to deserialize it by himself. Futhermore with this approach only the processor is called if he really wants to process this kind of event, before each processor had to check on each event if he want to process it.

# Prototype

The current protoype implementation can be found on https://github.com/zeebe-io/zeebe/tree/reactive-Stream-procs . I had to add some copy strategy to the message pack module, see this branch as well https://github.com/zeebe-io/zb-msgpack-json-path/tree/reactive-stream-processor .

The architecture of the reactive stream processor prototype could be summarized with the following picture:

![reactivestreamprocessor](https://user-images.githubusercontent.com/2758593/38050292-c8df210e-32ca-11e8-8a3a-6494f7b8805b.png)

**TODO** describe a bit further

**Comment by ThorbenLindhauer** [comment]
With the `TypedEventProcessor` and related interfaces we have achieved some of your ideas regarding the programming model, see for example `TaskInstanceStreamProcessor`: https://github.com/zeebe-io/zeebe/blob/master/broker-core/src/main/java/io/zeebe/broker/task/processor/TaskInstanceStreamProcessor.java. I'm sure we can make further simplifications. Your ideas regarding stream processor controller implementation sound very good.

**Comment by Zelldon** [comment]
Thanks :) Yeah this is cool. The TypedEventProcessor helped me a bit to implement the prototype.

**Comment by Zelldon** [comment]
Since you guys @saig0 @ThorbenLindhauer refactor the stream processors and impl new features, does it make sense to tackle/implement this as well or think about it?

**Comment by ThorbenLindhauer** [comment]
Some of this is already implemented for typed stream processors (e.g. automatically filtering records based on metadata to avoid deserializing records that are not relevant anyway).

Furthermore, we have the following issues:

* https://github.com/zeebe-io/zeebe/issues/1009: Single stream processor for jobs, wf instances, incidents. I considered doing this, but it wouldn't make our lives so much easier right now. Definitely something to do in the future, though
* https://github.com/zeebe-io/zeebe/issues/1048: No more distinction between StreamProcessor and EventProcessor. I tried to do this a couple of weeks ago, but the topic subscription stream processors were not easy to port. We can try again once we have removed those. For typed stream processors, we only program with a single interface anyway, so there the dual interfaces are hidden.

I have the impression that there are more ideas in your original post than what I covered now, but I'm not sure I understood everything. I encourage you to create individual tickets for the aspects I missed, so that we can discuss, track and prioritize them individually. 

In general, performance optimizations have currently rather low priority.
